src/index.js

```diff
@@ -56,6 +57,26 @@ module.exports = (web3, options = {}) => {
   const formatVersion = version =>
     version.split('.').map((part) => parseInt(part))
 
+  // https://gist.github.com/john-doherty/bcf35d39d8b30d01ae51ccdecf6c94f5
+  function promiseTimeout(ms, promise){
+    return new Promise(function(resolve, reject){
+      // create a timeout to reject promise if not resolved
+      var timer = setTimeout(function(){
+        reject(new Error("promise timeout"));
+      }, ms);
+
+      promise
+        .then(function(res){
+          clearTimeout(timer);
+          resolve(res);
+        })
+        .catch(function(err){
+          clearTimeout(timer);
+          reject(err);
+        });
+    });
+  };

```

Comment:
A little cleaner implementation, IMO:

```js
function promiseTimeout(promise, ms) {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error('promise timeout'))
    }, ms)

    promise
      .then(resolve)
      .catch(reject)
      .finally(() => clearTimeout(timer))
  }
}
```

Notes:

- `.finally()` is really useful for the duplicate logic
- `fn(promise, ms)` is more similar to other APIs, e.g. `setTimer()`, whose timing argument goes second.

However, the bit about using a `setTimeout` is kind of sketchy; it's better to use [`Promise.race()`](https://italonascimento.github.io/applying-a-timeout-to-your-promises/).

Reply:
Yes, I already saw that article (first Google entry for "javascript promise timeout", haha). Why do you think it's better? What do you mean by "sketchy"? What's the problem with `setTimeout`? The `Promise.race` solution is still using it. I'm changing it to that new way, but I would like to know the advantages.

Reply:
I think `Promise.race()` is conceptually cleaner, in a very FP-style of doing things.

The flow in the below:

```js
function timeoutPromise(ms) {
  return new Promise((resolve, reject) => {
    setTimeout(() => reject(new Error('time out')), ms)
  })
}

const yourPromise = <promise>
const timerPromise = timeoutPromise(<ms>)

Promise
  .race(yourPromise, timerPromise)
  .then(console.log) // yourPromise succeeded
  .catch(console.err) // generic error handler, could add specific logic in case yourPromise fails or it timeouts
```

I find the intents better separated and the control flow very clear. Rather than using the `setTimeout()` to short-circuit the top-layer promise, in this flow you use `setTimeout()` to create two separate "execution threads" joined by the `.race()`, which handles the merging of the threads into a single point.

I've probably drank too much of the koolaid, but this mental model comes from [railway oriented programming](https://vimeo.com/113707214).